README.md
imagepreview
============
[![Build Status](https://travis-ci.org/GaborWnuk/image-preview-python.svg?branch=master)](https://travis-ci.org/GaborWnuk/image-preview-python)

Simple helper module for Image Preview method for REST and GraphQL for immediate image preview on your client's side (Swift, Java, JavaScript and so on).

Idea is to deliver only around 200 bytes of image as a normal base64 data in one of JSON fields of your entities and then issue request to obtain full resolution of an image.

## Short "how does it work"
The main reason to do such thing is to show your user approximate look of your data stream as soon as possible. This method gives you the ability to serve image previews in the sam milisecond you show your text content in your app, as image preview is delivered as one of data fields in your JSON response, i.e:

```json
{
  "data": {
    "articles": [
      {
        "title": "Piotr Cyrwus zostanie fanatykiem wędkarstwa",
        "img": {
          "url": "https://i.wpimg.pl/1000-666/CSM//d.wpimg.pl/1636256103--1868262584_piotr-cyrwus.jpg",
          "b64": "AD8AqOc54pAMDNKTyRnrTkUS/JnBxUAOVSR8qE5pGVwcOu09qvD5AAo+XFVrkk7TjHNMdiLoB7UmT7UpHzU3A9TSEIynOaWLJlGDzmn3DL5zKigAHGR3otQGlxj+E00gLJb5euDUDtukxg8CrJTdVO6Y/aDjjaAOKdhsUZLDilbBY49aiSXYRkZFWMbud3XnrUtWEVCSWJPU81Zsf+Pg/wC7VX+KrVl/x8fhWiEaGBnNY8jb5Gb1Oa1pP9S/+6aypgBKwAwKGAw1cjdfLXI7DvVKlyfWpauM/9k="
        }
      }
    ]
  }
}
```

This library is sort of a helper to achieve above.

Because every thumbnail is compressed as JPEG, with the same size (default: 42x42 pixels) and using the same Huffman table, each JPEG file consist of a common header (620 bytes) and a small picture data part (around 200 bytes, depending on quality you use).

Having this in mind we choose to omit headers and send only 200 bytes (base64) and then prepend this data with a header on client side.

```python
file_bytes = open("matheson.jpg", "rb")

ip = ImagePreview(file_path_or_bytes=file_bytes)

# this is what you should have on your client side and prepend image_body with it
image_header = ip.thumbnail_header_bytes()

# this is what you send in your API to your client
image_body = ip.thumbnail_body_b64()
```

Or fetch from remote location:
```python
import requests

url = "https://upload.wikimedia.org/wikipedia/commons/4/44/Mt_Cook,_NZ.jpg"
response = requests.get(url)
ip = ImagePreview(file_path_or_bytes=BytesIO(response.content))

# this is what you send in your API to your client
image_body = ip.thumbnail_body_b64()
```

You can find basic Swift / iOS example here: [https://github.com/GaborWnuk/image-preview-ios-demo](https://github.com/GaborWnuk/image-preview-ios-demo).

## How to use
To install the package simply use PyPi
```
$ pip install imagepreview
```

## License
MIT License

Copyright (c) 2016 Gabor Wnuk

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGES
v0.0.1, Fri Oct 28 10:48:41 2016 -- Initial Release.


tests/imagepreview_test.py
# -*- coding: utf-8 -*-

"""
    imagepreview_test.py
    ~~~~~~~~~~~

    Unit tests for imagepreview package.

    :copyright: (c) 2016 by GaborWnuk.
    :license: see LICENSE for more details.
"""

import os
import sys
import unittest

path = os.path.dirname(os.path.abspath(__file__))
sys.path.insert(0, path + '/../')

from imagepreview import ImagePreview

class ImagePreviewTest(unittest.TestCase):
    def test_image_bytes_open(self):
        file_bytes = open("%s/assets/emerald.jpg" % path, "rb")
        ip = ImagePreview(file_path_or_bytes=file_bytes)

    def test_image_bytes_attribute_error(self):
        with self.assertRaises(AttributeError):
            ip = ImagePreview(file_path_or_bytes=0)

    def test_image_file_not_found(self):
        with self.assertRaises(IOError):
            ip = ImagePreview(file_path_or_bytes="Nonexistent file")

    def test_image_bytes_not_image(self):
        with self.assertRaises(IOError):
            file_bytes = open("%s/assets/bytes.bin" % path, "rb")
            ip = ImagePreview(file_path_or_bytes=file_bytes)

    def test_image_header_length(self):
        file_bytes = open("%s/assets/emerald.jpg" % path, "rb")
        ip = ImagePreview(file_path_or_bytes=file_bytes)

        self.assertEqual(len(ip.thumbnail_header_bytes()), 620)

    def test_image_header_body_length(self):
        file_bytes = open("%s/assets/emerald.jpg" % path, "rb")
        ip = ImagePreview(file_path_or_bytes=file_bytes)

        self.assertEqual(len(ip.thumbnail_body_bytes()), 226)

    def test_image_b64(self):
        file_bytes = open("%s/assets/emerald.jpg" % path, "rb")
        ip = ImagePreview(file_path_or_bytes=file_bytes)

        self.assertEqual(len(ip.thumbnail_b64()), 1128)

    def test_image_header_b64(self):
        file_bytes = open("%s/assets/emerald.jpg" % path, "rb")
        ip = ImagePreview(file_path_or_bytes=file_bytes)

        self.assertEqual(len(ip.thumbnail_header_b64()), 828)

    def test_image_body_b64(self):
        file_bytes = open("%s/assets/emerald.jpg" % path, "rb")
        ip = ImagePreview(file_path_or_bytes=file_bytes)

        self.assertEqual(len(ip.thumbnail_body_b64()), 304)

    def test_check_if_header_from_different_image_creates_proper_thumbnail(self):
        from PIL import Image
        from io import BytesIO

        file_bytes = open("%s/assets/emerald.jpg" % path, "rb")
        file_bytes_second = open("%s/assets/matheson.jpg" % path, "rb")

        ip = ImagePreview(file_path_or_bytes=file_bytes)
        ip_second = ImagePreview(file_path_or_bytes=file_bytes_second)

        composited_image_bytes = ip.thumbnail_header_bytes() + ip_second.thumbnail_body_bytes()

        composited_image_bytes = ip.thumbnail_bytes()

        ip_composited = Image.open(BytesIO(composited_image_bytes))

        width, height = ip_composited.size

        self.assertEqual(width, 42)
        self.assertEqual(height, 42)
        self.assertEqual(ip_composited.format, "JPEG")




imagepreview/__init__.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
    __init__.py
    ~~~~~~~~~~~

    no description available

    :copyright: (c) 2016 by GaborWnuk.
    :license: see LICENSE for more details.
"""

__title__ = 'imagepreview'
__version__ = '0.0.1'
__author__ = 'Gabor Wnuk'

import base64
from io import BytesIO
from PIL import Image, ImageOps

class ImagePreview(object):
    """Simple helper module for Image Preview method for REST and GraphQL for
    immediate image preview on your client's side (Swift, Java, JavaScript
    and so on).

    Idea is to deliver only around 200 bytes of image data as a normal
    base64 data in one of JSON fields of your entities and then issue request
    yo obtain full resolution of an image.

    Attributes:
        file_path_or_bytes: Standard PIL Image constructor argument for either
                            file path (string) or BytesIO object.
        thumbnail_size: desired thumbnail image size.
        quality: desired JPEG compression quality, where 0 is the lowest quality,
                    and 100 the highest.
    """

    __image = None

    __thumbnail_bytes = None
    __thumbnail_header_bytes = None
    __thumbnail_body_bytes = None

    __header_size = 620

    def __init__(self, file_path_or_bytes, thumbnail_size=(42, 42), quality=15):
        self.__image = Image.open(file_path_or_bytes)

        thumbnail = ImageOps.fit(self.__image, thumbnail_size, Image.ANTIALIAS)
        thumbnail_bytes = BytesIO()

        thumbnail.save(thumbnail_bytes, "JPEG", quality=quality, optimize=False)

        self.__thumbnail_bytes = thumbnail_bytes.getvalue()

        self.__thumbnail_header_bytes = self.__thumbnail_bytes[:self.__header_size]
        self.__thumbnail_body_bytes = self.__thumbnail_bytes[self.__header_size:]

    def thumbnail_bytes(self):
        return self.__thumbnail_bytes

    def thumbnail_header_bytes(self):
        return self.__thumbnail_header_bytes

    def thumbnail_body_bytes(self):
        return self.__thumbnail_body_bytes

    def thumbnail_b64(self):
        return base64.b64encode(self.__thumbnail_bytes).decode('utf-8')

    def thumbnail_header_b64(self):
        return base64.b64encode(self.__thumbnail_header_bytes).decode('utf-8')

    def thumbnail_body_b64(self):
        return base64.b64encode(self.__thumbnail_body_bytes).decode('utf-8')

